propose RFC process (meta proposal)

jzaefferer opened the pull request:
As discussed on Discord, I've copied parts of Rust's RFC process and adapted it to Nushell. Changing the template was pretty easy, changing the README was much much more work. Accordingly I expect that to need a much more thorough review.

Diffing against the original might help; I could add that as a gist later.

I noticed a few small issues in the original README and submitted that as a PR to their repo: https://github.com/rust-lang/rfcs/pull/2938

jzaefferer committed:
propose RFC process (meta proposal)

enesf commented on README.md:
```diff
@@ -0,0 +1,262 @@
+Many changes, including bug fixes and documentation improvements can be
+implemented and reviewed via the normal GitHub pull request workflow.
+
+Though ssome changes are substantial enough that we ask for these to be
```
```suggestion
Though some changes are substantial enough that we ask for these to be
```

enesf commented on README.md:
```diff
@@ -0,0 +1,262 @@
+
+In general, once accepted, RFCs should not be substantially changed. Only very
+minor changes should be submitted as amendments. More substantial changes
+should be new RFCs, with a note added to the original RFC. Exactly what counts
```
```suggestion
should be new RFCs, with a note added to the original RFC. What exactly counts
```

sophiajt commented on README.md:
```diff
@@ -0,0 +1,262 @@
+
+## Before creating an RFC
+[Before creating an RFC]: #before-creating-an-rfc
+
```
I feel like there's something the Nu team(s) should do to help the RFC process. This paragraph assumes that the person writing it understands what "hastily-proposed RFC" means.

One possible thing we should talk about is to help direct people towards mentors who can help refine the ideas so that they cover the necessary steps in a way that doesn't read as hasty, but instead fits better with the overall direction and style of RFCs.

Something else that I like are the separate stages. In the Rust RFC process, I felt like we would sometimes jump into talking about the details of a solution before we'd really agreed on the exact problem being solved. Making the stages clear, and having a way for people to understand how to move between the stages could also be helpful for both reviewers and writers.

jzaefferer replied:
> One possible thing we should talk about is to help direct people towards mentors [...]

Given the current state, effectively that would mean "chat with us on Discord in #design-discussion", wouldn't it?

If not, could you go into more detail?

> [stages]

Do you have something like the TC39 stages in mind? https://tc39.es/process-document/ At least I think those could be a good reference.

I don't think we should start with that, though. I'd rather leave it open to adopt stages like that later.

We can replace some of the language here, like "hastily" and make "strawperson" (from TC39) RFCs more welcome.

sophiajt replied:
Good points. Yes, agreed that "please come talk to us on discord" is probably the right first step :)

And right, I was thinking something like tc39 stages (I think ember uses something similar, too, though I'm less familiar with their process). Agreed that's too much for this stage. Right now just agreeing on the problem and having the RFC being a rough sketch of what they want to try should be plenty for us to use.

jzaefferer replied:
Looking at https://github.com/emberjs/rfcs I don't see well-defined stages. It looks very similar to Rust's process.

> Good points. Yes, agreed that "please come talk to us on discord" is probably the right first step :)

Its kinda in this section already, but a bit buried, and without mentioning the channel. I'll rework that.

jzaefferer replied:
I've dropped the entire first paragraph (no more mention of "hastily") and edited the others a bit.

sophiajt commented on README.md:
```diff
@@ -0,0 +1,262 @@
+feature as a pull request to the Nushell repo. Being "active" is not a rubber
+stamp, and in particular still does not mean the feature will ultimately be
+merged; it does mean that in principle all the major stakeholders have agreed
+to the feature and are amenable to merging it.
```
Something else that I noticed in the Rust RFC process is that people would resist an RFC getting accepted, even though its acceptance didn't mean that it would become part of the project. Rather, it just meant it could be implemented.

Going along with the idea above we might want to call out the explicit stages, we can call out that just because something has moved to being an experimental feature does not mean it will become part of the project. Instead, there is a review that happens that allows these experimental features to land in the project.

This is all assuming a post-1.0 view for Nu, so that every step is very intentional. Getting to 1.0, being agile is going to be important, so like you mention, we're going to want to be flexible.

jzaefferer replied:
That's good feedback. How about we lower the requirements for accepting an RFC, for now? That can still be adjusted later.

For example, we could do this:
```diff
- - Before actually entering FCP, *all* team members must sign off;
+ - Before actually entering FCP, *at least two* team members must sign off;
```

sophiajt replied:
@jzaefferer - that sounds good. We could even skip the FCP stage. If the proposal has gotten to the point where two team members sign off, it should be good enough to go ahead and implement.

jzaefferer replied:
Makes sense! I've dropped the entire FCP stage.  This simplifies 'What the process is' quite a bit. I've replaced it with two simpler variations:

> - When the RFC is in a state that justifies community attention, it should be
    advertised widely, e.g. in [This Week in Nushell](https://www.notion.so/This-Week-in-Nu-ed589e19f1be47fa9ef562c1757e7ba1).
    This way all stakeholders have a chance to lodge any final objections
    before a decision is reached.
> - At some point after advertising the RFC was advertised publicly, a member
    of the team will propose a decision for the RFC (merge, close, or postpone).
    Another team member must sign off on that decision, and can then take the
    appropriate action. When closing or postponing, the reasons need to be well
    documented in the PR.

This still provides a certain minimal duration, but without the FCP overhead.

sophiajt commented on README.md:
```diff
@@ -0,0 +1,262 @@
+is "no", then the appropriate response is to close the RFC, not postpone it.)
+
+
+### Help this is all too informal!
```
This section feels off to me (I see it's in the original Rust documentation, too). Let's say the person reading this feels like this is a ton of steps and get through and thinks "what? why would I think this is informal?".  It's confusing, at best, and unfortunately I think it might read as condescending depending on the reader and their background.

jzaefferer replied:
I agree with you, let's drop it.

jzaefferer replied:
Removed

jzaefferer committed:
readme: fix typo

Co-authored-by: Enes Fazli <fazli.enes@gmail.com>

jzaefferer committed:
readme: fix word order

Co-authored-by: Enes Fazli <fazli.enes@gmail.com>

jzaefferer committed:
readme: remove 'Help this is all too informal'

Not helpful, maybe condescending

jzaefferer committed:
readme: add reference to TC39 process for future stages

jzaefferer committed:
readme: rework 'Before creating an RFC'

Remove the gate-keeping language and instead focus on getting feedback up-front.

jzaefferer committed:
readme: remove fcp stage, use a simpler acceptance process

One team member suggests a decision, another signs off and takes action.

thegedge commented on README.md:
```diff
@@ -0,0 +1,229 @@
+
+  - [Opening](#nushell-rfcs)
+  - [Table of Contents]
+  - [When you need to follow this process]
```
need → should, until the note below is removed? Or will we change to "need to" before merging?

jzaefferer replied:
Good catch. I had changed the header, but missed this ToC entry. Fixed!

thegedge commented on README.md:
```diff
@@ -0,0 +1,229 @@
+    (warning removal, speedup, better platform coverage, more parallelism,
+    better handle errors, etc.)
+  - Additions only likely to be _noticed by_ other developers of Nushell,
+    invisible to users of Nushell.
```
The language above doesn't scream "these lists aren't necessarily comprehensive" to me. I wonder if we could somehow be more explicit about that?

jzaefferer replied:
I changed the line before the list to this: "Some changes do not require an RFC. For example:"

thegedge replied:
👍 "For example" implies non-comprehensiveness.

thegedge commented on README.md:
```diff
@@ -0,0 +1,229 @@
+    not change meaning".
+  - Additions that strictly improve objective, numerical quality criteria
+    (warning removal, speedup, better platform coverage, more parallelism,
+    better handle errors, etc.)
```
There's a conflict here. Speeding things up could mean a huge re-architecture that potentially changes behaviour. Should we emphasize a priority on the first list?

jzaefferer replied:
I'm don't know how to put this into words. Could you suggest an edit?

thegedge replied:
I'm actually not sure. I think a combination of the non-comprehensiveness of these lists plus the line about "what constitutes a substantial change is evolving" is probably good enough. I think any attempt to be more precise here will just end up making this process more difficult to understand 🙂

thegedge commented on README.md:
```diff
@@ -0,0 +1,229 @@
+
+If you are interested in working on the implementation for an "active" RFC, but
+cannot determine if someone else is already working on it, feel free to ask
+(e.g. by leaving a comment on the associated issue).
```
I've been making use of assignees on Github issues to indicate that someone is working on an issue. Do we want to take that approach on the associated Github issue to indicate someone is actively working on it? If yes, should we call that out here?

jzaefferer replied:
I added "; a team member can then assign you to the issue)."

thegedge commented on README.md:
```diff
@@ -0,0 +1,229 @@
+
+### Contributions
+
+Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.
```
This part doesn't make sense if we'll only license under MIT. Probably just need to drop reference to the Apache license and "dual licensing".

Or do we want to allow dual licensing? What are the benefits?

sophiajt replied:
For now, probably better to have everything under the same license as Nu, which should be MIT.

jzaefferer replied:
Yeah, somehow I missed that. I'll update it. I'd remove the section - saying that the license is MIT should be enough.

thegedge reviewed (commented):
Just a few small things, but this is great 🙌 🎉 🚀

jzaefferer committed:
readme: remove irrelevant contributions section

sophiajt commented on text/0001-establish-rfc-process.md:
```diff
@@ -0,0 +1,55 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The concept of the "team" is currently not well defined (to my knowledge). It doesn't exist in Rust, which instead has sub-teams (for compiler, language and libraries). Actually defining "team" for Nushell, for example as all commiters to the nushell/nushell repostiroy, would be a potential solution.
```
I'm okay defining the "Nu team" as all the committers. I'd be great to do a bit more here, but perhaps this can be defined in a follow-up RFC, since we could open things up to having a website team, data science team, etc

jzaefferer replied:
I added a definition in the Opening and replaced these two paragraph, since the question is now resolved.

sophiajt commented:
@thegedge @andrasio - any blockers or things you feel like we should add/change before moving this forward?

jzaefferer committed:
readme: small fixes, addressing thegedge's comments

jzaefferer committed:
add definition for "team", resolve that in the first rfc

thegedge commented:
Nope, this LGTM 🙌

andrasio commented:
Looks good. Frankly, I'd love to avoid stating it as RFC and call them `Nu Plays` instead.

jzaefferer commented:
> Frankly, I'd love to avoid stating it as RFC and call them Nu Plays instead.

I'm not sure how serious you are about this. I don't get the "Nu Plays" name, and I don't like the idea of the substantial rename work.

@jonathandturner do you want to feature this in the next "This week in Nushell" first? Or can it be merged now (and featured afterwards)?

sophiajt commented:
@jzaefferer - sorry for the slow response. Need to filter this new repo to the right place :).

Sounds like we can go ahead and land. I'll put it in the next This Week in Nu also.

andrasio commented:
> > Frankly, I'd love to avoid stating it as RFC and call them Nu Plays instead.
> 
> I'm not sure how serious you are about this. I don't get the "Nu Plays" name, and I don't like the idea of the substantial rename work.
> 
> @jonathandturner do you want to feature this in the next "This week in Nushell" first? Or can it be merged now (and featured afterwards)?

@jzaefferer Since the beginning we want to make sure Nu is enjoyable and fun. The sandbox for tests is called `playground` for instance (like kids having fun), users are `nuños`, in spanish `niños` means kids. `Nu Plays` as in a theater, I meant.

jzaefferer commented:
> I'll put it in the next This Week in Nu also.

Looks like you forgot :o

sophiajt commented:
@jzaefferer I realized that seconds after I posted it, oops. I went back and added it.
